File: src/services/document/paragraph-format.ts
/**
 * A formatação de parágrafo como o diálogo a mostra, e como o bloco a guarda.
 *
 * Os atributos já existem no modelo e o gravador já sabe escrevê-los (ver
 * `extensions/block-format.ts` e `ParagraphFormat.cs`): o que faltava era a
 * tradução entre o que o arquivo representa e o que uma pessoa consegue
 * preencher num formulário. As duas formas não são a mesma:
 *
 *  - a entrelinha mora num campo só, como o CSS a escreve — `normal`, um número
 *    (`1.8311`) ou uma medida (`14pt`) —, e no diálogo são duas perguntas: que
 *    tipo de espaçamento, e quanto. E o número do CSS **não** é o fator do Word:
 *    é ele já multiplicado pela altura natural da fonte (ver `line-metrics.ts`).
 *    Tratar um pelo outro é o que fazia todo parágrafo de Calibri abrir como
 *    "Múltiplo 1,2207" e "1,5" gravar 1,23 linha no arquivo;
 *  - o recuo de primeira linha é **um** número com sinal, porque no CSS
 *    `text-indent` negativo é o deslocamento do Word; no diálogo são a escolha
 *    ("nenhum", "primeira linha", "deslocamento") e uma medida positiva.
 *
 * Aqui só a conversão e os limites. Nada de React, nada de Tiptap: é o que
 * permite testar as duas direções sem montar editor nenhum.
 */

import { cssLineHeightOf, explicitCssLineHeightOf, lineFactorOf } from './line-metrics.js'

/** Como a entrelinha é medida. Os três casos que o gravador sabe escrever. */
export const LineSpacingKind = {
  /** A altura que a própria fonte pede — o silêncio do arquivo. */
  Single: 'single',
  /** Múltiplo da altura natural: 1,15, 1,5, duplo. */
  Multiple: 'multiple',
  /**
   * Medida fixa em pontos.
   *
   * "Pelo menos", e não "exatamente": é assim que o gravador a escreve
   * (`w:lineRule="atLeast"`), porque `exact` corta o que não couber na altura
   * declarada — e perder metade de uma linha é perda de conteúdo.
   */
  AtLeast: 'at-least',
} as const

export type LineSpacingKind = (typeof LineSpacingKind)[keyof typeof LineSpacingKind]

/** O que o recuo da primeira linha faz: nada, entra ou sai. */
export const FirstLineKind = {
  None: 'none',
  Indent: 'indent',
  Hanging: 'hanging',
} as const

export type FirstLineKind = (typeof FirstLineKind)[keyof typeof FirstLineKind]

export const TextAlignment = {
  Left: 'left',
  Center: 'center',
  Right: 'right',
  Justify: 'justify',
} as const

export type TextAlignment = (typeof TextAlignment)[keyof typeof TextAlignment]

/** O formulário inteiro, já em números — nunca em texto de `<input>`. */
export interface ParagraphDraft {
  readonly align: TextAlignment
  /** Pontos, como o Word os mostra. */
  readonly spaceBefore: number
  readonly spaceAfter: number
  readonly lineSpacingKind: LineSpacingKind
  /** Fator quando o tipo é múltiplo; pontos quando é "pelo menos". */
  readonly lineSpacingValue: number
  /** Milímetros, como a régua da configuração de página. */
  readonly indentLeftMm: number
  readonly indentRightMm: number
  readonly firstLineKind: FirstLineKind
  readonly firstLineMm: number
  readonly keepNext: boolean
}

/** Meia polegada, em milímetros: o passo de recuo do OOXML (720 twips). */
export const INDENT_STEP_MM = 12.7

export const MAX_SPACING_PT = 600
export const MAX_INDENT_MM = 200
/**
 * Os limites do fator, e por que são estes.
 *
 * O gravador recusa fator fora de `(0,5; 4)` e registra perda — ver
 * `ParagraphFormat.ApplyLineHeight`. Oferecer no diálogo o que ele não escreve
 * seria prometer o que não se cumpre.
 *
 * São limites do fator **do Word**, e é a conversão de `line-metrics.ts` que os
 * torna verdadeiros: enquanto o número do diálogo ia cru para o atributo, 0,51
 * em Calibri chegava ao gravador como 0,42 — fora da faixa, perda no inventário
 * e nada escrito no arquivo.
 */
export const MIN_LINE_FACTOR = 0.51
export const MAX_LINE_FACTOR = 3.99

export const DEFAULT_PARAGRAPH_DRAFT: ParagraphDraft = {
  align: TextAlignment.Left,
  spaceBefore: 0,
  spaceAfter: 0,
  lineSpacingKind: LineSpacingKind.Single,
  lineSpacingValue: 1.15,
  indentLeftMm: 0,
  indentRightMm: 0,
  firstLineKind: FirstLineKind.None,
  firstLineMm: 0,
  keepNext: false,
}

/** Os atributos que o diálogo escreve no bloco. Valor nulo apaga o atributo. */
export interface ParagraphAttrs {
  /**
   * Nulo quando ninguém escolheu alinhamento.
   *
   * Pela mesma razão dos recuos: o leitor omite o atributo no parágrafo sem
   * `w:jc`, e escrever `left` ali acrescentaria um `w:jc` que o documento não
   * tinha — bloco reescrito à custa de um "Aplicar" que não mudou nada.
   */
  readonly textAlign: TextAlignment | null
  readonly spaceBefore: number
  readonly spaceAfter: number
  readonly lineHeight: string
  readonly indentMm: number | null
  readonly indentRightMm: number | null
  readonly firstLineMm: number | null
  readonly keepNext: true | null
  /**
   * O recuo em passos de `Ctrl+]` é zerado.
   *
   * As duas origens somam no gravador (medida do arquivo + nível do editor), e
   * o diálogo fala em milímetros: deixar o nível de pé faria o campo mostrar
   * 10 mm e o arquivo receber 10 mm mais dois passos.
   */
  readonly indent: 0
}

const clamp = (value: number, min: number, max: number): number =>
  Number.isFinite(value) ? Math.min(Math.max(value, min), max) : min

/** Arredonda para o décimo: é a precisão com que o leitor entrega a medida. */
const round = (value: number): number => Math.round(value * 10) / 10

function numberOf(value: unknown): number | null {
  if (value === null || value === undefined || value === '') return null
  const parsed = Number(value)
  return Number.isFinite(parsed) ? parsed : null
}

/**
 * O bloco selecionado → o formulário.
 *
 * Recebe os atributos crus do nó porque é assim que o Tiptap os entrega, e
 * porque um bloco importado traz medidas que nenhum campo do diálogo criou.
 */
export function paragraphDraftFrom(attrs: Record<string, unknown>): ParagraphDraft {
  const firstLine = numberOf(attrs['firstLineMm']) ?? 0
  const indentLevel = numberOf(attrs['indent']) ?? 0

  return {
    align: alignmentOf(attrs['textAlign']),
    spaceBefore: clamp(round(numberOf(attrs['spaceBefore']) ?? 0), 0, MAX_SPACING_PT),
    spaceAfter: clamp(round(numberOf(attrs['spaceAfter']) ?? 0), 0, MAX_SPACING_PT),
    ...lineSpacingOf(attrs['lineHeight'], fontStackOf(attrs)),
    // O nível de `Ctrl+]` entra como medida, senão o recuo que a pessoa vê na
    // tela não é o que o campo mostra. 12,7 mm é o passo de meia polegada que o
    // OOXML usa (720 twips), o mesmo de `TwipsPerIndentLevel` no gravador.
    indentLeftMm: clamp(
      round((numberOf(attrs['indentMm']) ?? 0) + indentLevel * INDENT_STEP_MM),
      0,
      MAX_INDENT_MM,
    ),
    indentRightMm: clamp(round(numberOf(attrs['indentRightMm']) ?? 0), 0, MAX_INDENT_MM),
    firstLineKind:
      firstLine > 0 ? FirstLineKind.Indent : firstLine < 0 ? FirstLineKind.Hanging : FirstLineKind.None,
    firstLineMm: clamp(round(Math.abs(firstLine)), 0, MAX_INDENT_MM),
    keepNext: attrs['keepNext'] === true,
  }
}

function alignmentOf(value: unknown): TextAlignment {
  const found = Object.values(TextAlignment).find((align) => align === value)
  return found ?? TextAlignment.Left
}

/** A pilha de fontes do bloco, que é quem diz a altura natural da linha. */
function fontStackOf(attrs: Record<string, unknown>): string | null {
  const stack = attrs['fontFamily']
  return typeof stack === 'string' && stack.trim() !== '' ? stack : null
}

function lineSpacingOf(
  value: unknown,
  fontStack: string | null,
): Pick<ParagraphDraft, 'lineSpacingKind' | 'lineSpacingValue'> {
  if (typeof value !== 'string' || value === '' || value.toLowerCase() === 'normal') {
    return {
      lineSpacingKind: LineSpacingKind.Single,
      lineSpacingValue: DEFAULT_PARAGRAPH_DRAFT.lineSpacingValue,
    }
  }

  if (value.toLowerCase().endsWith('pt')) {
    const points = numberOf(value.slice(0, -2))
    return points === null || points <= 0
      ? {
          lineSpacingKind: LineSpacingKind.Single,
          lineSpacingValue: DEFAULT_PARAGRAPH_DRAFT.lineSpacingValue,
        }
      : {
          lineSpacingKind: LineSpacingKind.AtLeast,
          lineSpacingValue: clamp(round(points), 1, MAX_SPACING_PT),
        }
  }

  const css = numberOf(value)
  // O número do atributo é medida de CSS; o do diálogo é linha do Word. Um
  // parágrafo de Calibri sem entrelinha declarada chega aqui como 1,2207, e o que
  // a pessoa precisa ver é "Simples".
  const factor = css === null || css <= 0 ? 1 : lineFactorOf(css, fontStack)

  return factor === 1
    ? {
        lineSpacingKind: LineSpacingKind.Single,
        lineSpacingValue: DEFAULT_PARAGRAPH_DRAFT.lineSpacingValue,
      }
    : {
        lineSpacingKind: LineSpacingKind.Multiple,
        lineSpacingValue: clamp(factor, MIN_LINE_FACTOR, MAX_LINE_FACTOR),
      }
}

/**
 * O formulário → os atributos do bloco.
 *
 * Zero vira `null` nos recuos de propósito: ausência é o que o leitor produz
 * para o parágrafo sem recuo, e um `0` explícito faria a impressão digital ver
 * diferença onde não há — todo bloco seria reescrito ao salvar.
 */
export function paragraphAttrsFrom(
  draft: ParagraphDraft,
  attrs: Record<string, unknown> = {},
): ParagraphAttrs {
  const firstLine =
    draft.firstLineKind === FirstLineKind.None
      ? 0
      : draft.firstLineKind === FirstLineKind.Hanging
        ? -Math.abs(draft.firstLineMm)
        : Math.abs(draft.firstLineMm)

  return {
    textAlign: keptAlignment(draft.align, attrs),
    spaceBefore: clamp(round(draft.spaceBefore), 0, MAX_SPACING_PT),
    spaceAfter: clamp(round(draft.spaceAfter), 0, MAX_SPACING_PT),
    lineHeight: lineHeightOf(draft, attrs),
    indentMm: draft.indentLeftMm > 0 ? clamp(round(draft.indentLeftMm), 0, MAX_INDENT_MM) : null,
    indentRightMm: draft.indentRightMm > 0 ? clamp(round(draft.indentRightMm), 0, MAX_INDENT_MM) : null,
    firstLineMm: firstLine === 0 ? null : clamp(round(firstLine), -MAX_INDENT_MM, MAX_INDENT_MM),
    keepNext: draft.keepNext ? true : null,
    indent: 0,
  }
}

/**
 * O alinhamento, escrito só quando é escolha.
 *
 * "À esquerda" é o que o formulário mostra para o parágrafo que não declara
 * alinhamento nenhum — então, se era isso que estava lá, continua ausente. Num
 * parágrafo que **declara** outro alinhamento, escolher esquerda é uma decisão, e
 * apagar o atributo deixaria o estilo justificar de volta.
 */
function keptAlignment(align: TextAlignment, attrs: Record<string, unknown>): TextAlignment | null {
  const declared = typeof attrs['textAlign'] === 'string' ? attrs['textAlign'] : null
  return align === TextAlignment.Left && declared === null ? null : align
}

/**
 * A entrelinha como o CSS a escreve — a mesma forma que o leitor produz.
 *
 * Recebe os atributos atuais porque a conversão depende de duas coisas do bloco:
 * a fonte, que dá a altura natural da linha, e o que já estava escrito ali. O
 * segundo importa só no espaçamento simples, e os três caminhos estão comentados
 * abaixo.
 */
function lineHeightOf(draft: ParagraphDraft, attrs: Record<string, unknown>): string {
  if (draft.lineSpacingKind === LineSpacingKind.AtLeast) {
    return `${clamp(round(draft.lineSpacingValue), 1, MAX_SPACING_PT)}pt`
  }

  const fontStack = fontStackOf(attrs)

  if (draft.lineSpacingKind === LineSpacingKind.Single) {
    const current = attrs['lineHeight']

    // Nada declarado, ou `normal`: o simples do arquivo é o silêncio dele, e quem
    // mede melhor aí é o navegador. Trocar por número reescreveria o bloco sem
    // mudar uma linha do que se vê.
    if (current === null || current === undefined || current === '') return 'normal'
    if (typeof current === 'string' && current.toLowerCase() === 'normal') return 'normal'

    // Já era o fator 1, na forma em que o leitor o escreveu: fica como está, pelo
    // mesmo motivo.
    const css = numberOf(current)
    if (css !== null && css > 0 && lineFactorOf(css, fontStack) === 1) return String(current)

    // Havia medida declarada, e agora a escolha é simples. Aqui **não** vale
    // `normal`: é o único valor que o gravador não grava, e o `w:line` antigo
    // ficaria de pé — a entrelinha da pessoa perdida em silêncio.
    return explicitCssLineHeightOf(1, fontStack)
  }

  return cssLineHeightOf(clamp(draft.lineSpacingValue, MIN_LINE_FACTOR, MAX_LINE_FACTOR), fontStack)
}

/**
 * A entrelinha do bloco como os seletores rápidos a falam: `''` para simples, o
 * fator do Word (`1.5`) ou a medida em pontos (`14pt`).
 *
 * Existe para a barra de ferramentas e para os atalhos `Ctrl+1`, `Ctrl+5` e
 * `Ctrl+2`: eles falam em linha, como o Word, e o atributo fala em CSS.
 */
export function lineSpacingChoiceOf(attrs: Record<string, unknown>): string {
  const spacing = lineSpacingOf(attrs['lineHeight'], fontStackOf(attrs))

  if (spacing.lineSpacingKind === LineSpacingKind.Single) return ''
  if (spacing.lineSpacingKind === LineSpacingKind.AtLeast) return `${spacing.lineSpacingValue}pt`
  return String(spacing.lineSpacingValue)
}

/** O inverso: a escolha do seletor rápido → o atributo do bloco. */
export function lineHeightAttrFrom(choice: string, attrs: Record<string, unknown>): string {
  const spacing = choice.endsWith('pt')
    ? { lineSpacingKind: LineSpacingKind.AtLeast, lineSpacingValue: numberOf(choice.slice(0, -2)) ?? 12 }
    : choice.trim() === ''
      ? { lineSpacingKind: LineSpacingKind.Single, lineSpacingValue: 1 }
      : { lineSpacingKind: LineSpacingKind.Multiple, lineSpacingValue: numberOf(choice) ?? 1 }

  return lineHeightOf({ ...DEFAULT_PARAGRAPH_DRAFT, ...spacing }, attrs)
}

/**
 * O formulário está preenchido de um jeito que dá para aplicar?
 *
 * Só o que um número fora de faixa causaria: o resto dos campos é escolha
 * fechada, e o `clamp` já protege a conversão.
 */
export function isValidParagraphDraft(draft: ParagraphDraft): boolean {
  const inRange = (value: number, min: number, max: number): boolean =>
    Number.isFinite(value) && value >= min && value <= max

  if (!inRange(draft.spaceBefore, 0, MAX_SPACING_PT)) return false
  if (!inRange(draft.spaceAfter, 0, MAX_SPACING_PT)) return false
  if (!inRange(draft.indentLeftMm, 0, MAX_INDENT_MM)) return false
  if (!inRange(draft.indentRightMm, 0, MAX_INDENT_MM)) return false
  if (!inRange(draft.firstLineMm, 0, MAX_INDENT_MM)) return false

  if (draft.lineSpacingKind === LineSpacingKind.Multiple) {
    return inRange(draft.lineSpacingValue, MIN_LINE_FACTOR, MAX_LINE_FACTOR)
  }
  if (draft.lineSpacingKind === LineSpacingKind.AtLeast) {
    return inRange(draft.lineSpacingValue, 1, MAX_SPACING_PT)
  }

  return true
}

File: sidecar/src/Librevia.Format/Docx/ParagraphFormat.cs
using System.Globalization;
using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Wordprocessing;

namespace Librevia.Format.Docx;

/// <summary>
/// O `w:pPr` de um parágrafo editado: o original, com o que o modelo diz por
/// cima.
/// </summary>
/// <remarks>
/// Antes daqui o `w:pPr` era montado do zero a cada bloco reescrito, e com ele
/// iam embora o estilo do parágrafo, o espaçamento, a entrelinha, o fundo, o
/// "manter com o próximo", a fonte da marca de parágrafo e — pior de tudo — o
/// `w:sectPr` que fecha uma seção. O leitor lê todas essas coisas e o editor as
/// mostra na tela; o escritor as apagava em silêncio ao gravar. Corrigir uma
/// vírgula num título custava o título.
///
/// A regra é a mesma da gravação cirúrgica, um nível abaixo: **parte-se do XML
/// original** e só se sobrepõe o que o modelo de fato representa. O que o editor
/// não conhece — `w:pBdr`, `w:framePr`, `w:tabs`, `w:sectPr`, a marca de
/// parágrafo inteira — atravessa a edição intacto porque ninguém o reescreve.
/// </remarks>
internal sealed class ParagraphFormat(Inventory inventory)
{
    private const int TwipsPerIndentLevel = 720;

    /// <summary>
    /// O `w:pPr` a gravar, ou <c>null</c> quando não há nada a dizer.
    /// </summary>
    /// <param name="list">
    /// O que quem grava sabe sobre a numeração: <c>null</c> quando não sabe nada,
    /// e aí o `w:numPr` do original fica onde está.
    /// </param>
    /// <param name="original">O parágrafo como estava no arquivo, quando existe.</param>
    public ParagraphProperties? Build(Node node, ParagraphWriter.ListPlacement? list, Paragraph? original)
    {
        var properties = original?.ParagraphProperties?.CloneNode(true) as ParagraphProperties
                         ?? new ParagraphProperties();

        ApplyStyle(properties, node);
        ApplyAlignment(properties, node);
        ApplyIndentation(properties, node);
        ApplySpacing(properties, node);
        ApplyShading(properties, node);
        ApplyKeepNext(properties, node);
        ApplyMark(properties, node);

        // A numeração vem do contexto, e não do nó: no arquivo a lista são
        // parágrafos irmãos apontando o mesmo `w:numId`.
        //
        // Quem grava dentro de uma célula ou de uma caixa de texto não passa
        // contexto nenhum — `null` —, e isso não quer dizer "não é lista": quer
        // dizer "não sei". Aí o `w:numPr` do original fica de pé. Com embrulho,
        // a afirmação é do corpo: vazio por dentro é "deixou de ser item", e o
        // apontamento sai, senão o parágrafo continua numerado depois de a pessoa
        // ter tirado a lista.
        if (list is { } placement)
        {
            properties.NumberingProperties = placement.List is not { } context
                ? null
                : new NumberingProperties(
                    new NumberingLevelReference { Val = context.Level },
                    new NumberingId { Val = context.NumberingId });
        }

        return properties.HasChildren ? properties : null;
    }

    /// <summary>
    /// O estilo do parágrafo: o que o modelo carrega, e não um nome inventado.
    /// </summary>
    /// <remarks>
    /// O `styleId` viaja no modelo justamente para isto. Recalculá-lo a partir do
    /// nível do título trocava `Ttulo1` — como o LibreOffice o grava — por
    /// `Heading1`, que o documento não define: o título perdia a formatação
    /// inteira ao ser editado.
    ///
    /// O nível só manda quando o estilo que o modelo traz não é de título: é o
    /// caso do parágrafo que a pessoa transformou em título aqui dentro, e aí não
    /// há estilo original a respeitar.
    /// </remarks>
    private static void ApplyStyle(ParagraphProperties properties, Node node)
    {
        var declared = Attr.String(node, "styleId");
        var level = node.Type == "heading" ? Attr.Int(node, "level") : null;

        if (level is not null)
        {
            var keep = declared is not null && BodyReader.HeadingLevelOfStyle(declared) == level;
            properties.ParagraphStyleId = new ParagraphStyleId { Val = keep ? declared : "Heading" + level };
            return;
        }

        // Título que deixou de ser título não pode continuar apontando o estilo
        // de título: na tela virou parágrafo, e no arquivo continuaria barra
        // vermelha.
        if (declared is null || BodyReader.HeadingLevelOfStyle(declared) is not null)
        {
            if (declared is not null) properties.ParagraphStyleId = null;
            return;
        }

        properties.ParagraphStyleId = new ParagraphStyleId { Val = declared };
    }

    private static void ApplyAlignment(ParagraphProperties properties, Node node)
    {
        if (Attr.String(node, "textAlign") is not { } align) return;

        properties.Justification = new Justification
        {
            Val = align switch
            {
                "center" => JustificationValues.Center,
                "right" => JustificationValues.Right,
                "justify" => JustificationValues.Both,
                _ => JustificationValues.Left,
            },
        };
    }

    /// <remarks>
    /// A medida do arquivo primeiro, e o nível do editor por cima: são as duas
    /// origens do recuo, e o nível existe porque `Ctrl+]` trabalha em passos.
    /// Somá-los é o que faz recuar um parágrafo importado acrescentar um passo ao
    /// recuo que ele já tinha, em vez de apagá-lo.
    ///
    /// Modelo sem recuo nenhum é **afirmação**, e não silêncio: é o que volta de
    /// quem apertou `Ctrl+[` até o fim. Enquanto este método saía calado nesse
    /// caso, a diminuição não chegava ao arquivo e o recuo reaparecia ao reabrir
    /// o documento.
    ///
    /// Zera só o que o leitor de fato mostra na tela. Um recuo **negativo** — a
    /// primeira linha que sai para fora da margem — nunca chega ao editor, porque
    /// o leitor só emite a medida positiva; o zero que volta do modelo não está
    /// falando dele, e por isso ele fica.
    /// </remarks>
    private static void ApplyIndentation(ParagraphProperties properties, Node node)
    {
        var left = (Attr.MmToTwips(Attr.Double(node, "indentMm")) ?? 0)
                   + ((Attr.Int(node, "indent") ?? 0) * TwipsPerIndentLevel);
        var right = Attr.MmToTwips(Attr.Double(node, "indentRightMm")) ?? 0;
        var firstLine = Attr.MmToTwips(Attr.Double(node, "firstLineMm")) ?? 0;

        if (left <= 0 && right <= 0 && firstLine == 0)
        {
            ClearIndentation(properties.Indentation);
            return;
        }

        var indentation = properties.Indentation;
        if (indentation is null)
        {
            indentation = new Indentation();
            properties.Indentation = indentation;
        }

        // `Twips(...)`, e não o ternário direto: `x > 0 ? Invariant(x) : null`
        // tem tipo `string`, e o `null` de um `string` vira `StringValue(null)` na
        // conversão implícita — o SDK então grava `w:right=""` em vez de omitir o
        // atributo. O LibreOffice tolera; o Word recusa o documento inteiro.
        indentation.Left = Twips(left > 0 ? left : null);
        indentation.Right = Twips(right > 0 ? right : null);

        // Um só atributo, com o sinal decidindo qual: o Word grava `w:firstLine`
        // e `w:hanging` como dois, e declarar os dois deixaria o arquivo dizendo
        // duas coisas sobre a mesma linha.
        indentation.FirstLine = Twips(firstLine > 0 ? firstLine : null);
        indentation.Hanging = Twips(firstLine < 0 ? -firstLine : null);
    }

    /// <summary>O recuo positivo do arquivo, zerado — o negativo fica de pé.</summary>
    private static void ClearIndentation(Indentation? indentation)
    {
        if (indentation is null) return;

        // Zero explícito, e não a ausência do atributo: o estilo do parágrafo
        // também declara recuo, e apagar o atributo traria o do estilo de volta.
        if (IsPositive(indentation.Left)) indentation.Left = "0";
        if (IsPositive(indentation.Right)) indentation.Right = "0";

        // `w:firstLine` e `w:hanging` empurram e puxam a mesma linha, e nenhum é
        // o padrão do outro: aqui zerar é remover.
        if (IsPositive(indentation.FirstLine)) indentation.FirstLine = null;
        if (IsPositive(indentation.Hanging)) indentation.Hanging = null;
    }

    /// <summary>
    /// A medida do arquivo é positiva?
    /// </summary>
    /// <remarks>
    /// O que não é número inteiro de twips — o OOXML moderno aceita `0.5in` —
    /// devolve falso e fica onde está: não sabemos comparar, e mexer no que não
    /// se entende é como a medida do autor se perde.
    /// </remarks>
    private static bool IsPositive(StringValue? measure) =>
        int.TryParse(measure?.Value, NumberStyles.Integer, CultureInfo.InvariantCulture, out var twips)
        && twips > 0;

    private void ApplySpacing(ParagraphProperties properties, Node node)
    {
        var before = Attr.Double(node, "spaceBefore");
        var after = Attr.Double(node, "spaceAfter");
        var lineHeight = Attr.String(node, "lineHeight");
        if (before is null && after is null && lineHeight is null) return;

        var spacing = properties.SpacingBetweenLines;
        if (spacing is null)
        {
            spacing = new SpacingBetweenLines();
            properties.SpacingBetweenLines = spacing;
        }

        // Ponto → twip: vinte avos, e é exato nas duas direções até o décimo de
        // ponto, que é a precisão com que o leitor entrega a medida.
        if (before is not null) spacing.Before = Invariant((int)Math.Round(before.Value * 20));
        if (after is not null) spacing.After = Invariant((int)Math.Round(after.Value * 20));
        if (lineHeight is not null) ApplyLineHeight(spacing, lineHeight, Attr.String(node, "fontFamily"));
    }

    /// <summary>
    /// A entrelinha do CSS de volta para `w:line`.
    /// </summary>
    /// <remarks>
    /// O caminho de ida está em <c>BodyReader.LineHeightOf</c>, e este é o
    /// inverso dele. O múltiplo do OOXML vem em 240-avos e é medido sobre a
    /// **altura natural da fonte**, não sobre o tamanho dela — por isso a conta
    /// precisa saber qual fonte o parágrafo usa, que é a mesma informação que a
    /// leitura usou para produzir o número.
    ///
    /// `normal` não é convertido: quer dizer "a altura que a fonte pedir", que é
    /// justamente o silêncio do arquivo. Sobrescrever o `w:line` original com um
    /// palpite de 240 mudaria a paginação de um parágrafo que ninguém tocou na
    /// entrelinha.
    /// </remarks>
    private void ApplyLineHeight(SpacingBetweenLines spacing, string value, string? fontStack)
    {
        if (value.Equals("normal", StringComparison.OrdinalIgnoreCase)) return;

        if (value.EndsWith("pt", StringComparison.OrdinalIgnoreCase))
        {
            if (Attr.Points(value) is not { } points || points <= 0) return;

            spacing.Line = Invariant((int)Math.Round(points * 20));

            // Travada em "pelo menos", e não em "exatamente": as duas voltam do
            // leitor como pontos, e a que preserva conteúdo é esta — `exact`
            // corta o que não couber na altura declarada.
            if (spacing.LineRule is null || spacing.LineRule.Value == LineSpacingRuleValues.Auto)
            {
                spacing.LineRule = LineSpacingRuleValues.AtLeast;
            }

            return;
        }

        if (!double.TryParse(value, NumberStyles.Float, CultureInfo.InvariantCulture, out var css) || css <= 0)
        {
            inventory.NoteLoss($"entrelinha \"{value}\"");
            return;
        }

        var natural = LineMetrics.Of(FirstFont(fontStack)) ?? 1.1499;
        var factor = css / natural;
        if (factor is <= 0.5 or >= 4)
        {
            inventory.NoteLoss($"entrelinha \"{value}\"");
            return;
        }

        spacing.Line = Invariant((int)Math.Round(factor * 240));
        spacing.LineRule = LineSpacingRuleValues.Auto;
    }

    private void ApplyShading(ParagraphProperties properties, Node node)
    {
        if (Attr.String(node, "background") is not { } background) return;

        if (ColorValue.Hex(background) is not { } fill)
        {
            inventory.NoteLoss($"cor de fundo \"{background}\"");
            return;
        }

        properties.Shading = new Shading { Val = ShadingPatternValues.Clear, Color = "auto", Fill = fill };
    }

    /// <remarks>
    /// Ausente quer dizer desligado, e desligar é apagar o elemento — mas só
    /// quando ele estava ligado. Um `w:keepNext w:val="false"` existe para
    /// **desligar** o que o estilo liga, e removê-lo faria o parágrafo voltar a
    /// grudar no seguinte.
    /// </remarks>
    private static void ApplyKeepNext(ParagraphProperties properties, Node node)
    {
        if (Attr.Bool(node, "keepNext"))
        {
            properties.KeepNext = new KeepNext();
            return;
        }

        if (RunReader.IsOn(properties.KeepNext)) properties.KeepNext = null;
    }

    /// <summary>
    /// A fonte da marca de parágrafo (`w:pPr/w:rPr`).
    /// </summary>
    /// <remarks>
    /// É ela que o Word usa para medir a linha e para dar altura ao parágrafo
    /// vazio, e é de lá que o leitor tira a fonte do bloco. O resto da marca —
    /// negrito, cor, idioma — segue intocado no clone do original.
    /// </remarks>
    private void ApplyMark(ParagraphProperties properties, Node node)
    {
        var font = Attr.String(node, "fontFamily");
        var size = Attr.String(node, "fontSize");
        if (font is null && size is null) return;

        var mark = properties.ParagraphMarkRunProperties;
        if (mark is null)
        {
            mark = new ParagraphMarkRunProperties();
            properties.ParagraphMarkRunProperties = mark;
        }

        if (FirstFont(font) is { } family)
        {
            var fonts = mark.GetFirstChild<RunFonts>();
            if (fonts is null) PutInOrder(mark, new RunFonts { Ascii = family, HighAnsi = family });
            else
            {
                fonts.Ascii = family;
                fonts.HighAnsi = family;
            }
        }

        if (size is null) return;

        if (Attr.Points(size) is { } points && points > 0)
        {
            var halfPoints = Invariant((int)Math.Round(points * 2));
            var declared = mark.GetFirstChild<FontSize>();
            if (declared is null) PutInOrder(mark, new FontSize { Val = halfPoints });
            else declared.Val = halfPoints;
        }
        else
        {
            inventory.NoteLoss($"tamanho de fonte \"{size}\"");
        }
    }

    /// <summary>
    /// A ordem em que o `w:rPr` declara os seus filhos.
    /// </summary>
    /// <remarks>
    /// O OOXML é uma sequência, não um conjunto: `w:sz` depois de `w:u` torna o
    /// documento inválido, e o Word o recusa inteiro. A marca de parágrafo é a
    /// única propriedade que a biblioteca não expõe tipada — nas outras o próprio
    /// SDK põe cada elemento no lugar —, e por isso a ordem dela mora aqui.
    /// </remarks>
    private static readonly string[] MarkOrder =
    [
        "ins", "del", "moveFrom", "moveTo", "rStyle", "rFonts", "b", "bCs", "i", "iCs", "caps", "smallCaps",
        "strike", "dstrike", "outline", "shadow", "emboss", "imprint", "noProof", "snapToGrid", "vanish",
        "webHidden", "color", "spacing", "w", "kern", "position", "sz", "szCs", "highlight", "u", "effect",
        "bdr", "shd", "fitText", "vertAlign", "rtl", "cs", "em", "lang", "eastAsianLayout", "specVanish",
        "oMath", "rPrChange",
    ];

    private static void PutInOrder(ParagraphMarkRunProperties mark, OpenXmlElement child)
    {
        var rank = Array.IndexOf(MarkOrder, child.LocalName);

        // Elemento que não está na lista vai para o fim: é o que a extensão de
        // um fornecedor faz, e ela sempre se declara depois do padrão.
        var next = mark.ChildElements.FirstOrDefault(existing =>
        {
            var other = Array.IndexOf(MarkOrder, existing.LocalName);
            return other < 0 || other > rank;
        });

        if (next is null) mark.AppendChild(child);
        else mark.InsertBefore(child, next);
    }

    /// <summary>
    /// A primeira fonte da pilha do CSS.
    /// </summary>
    /// <remarks>
    /// O que sai do leitor é uma pilha — a fonte pedida e a substituta genérica —
    /// e o `w:rFonts` guarda o nome de uma fonte, não uma pilha.
    /// </remarks>
    internal static string? FirstFont(string? stack)
    {
        if (stack is null) return null;
        var first = stack.Split(',')[0].Trim().Trim('\'', '"');
        return first.Length > 0 ? first : null;
    }

    /// <summary>
    /// O número como o atributo o escreve, sem a vírgula de nenhuma região.
    /// </summary>
    /// <remarks>
    /// Chamava-se `Twips`, e mentia: metade dos usos é em meios-pontos (`w:sz`) e
    /// em 240-avos (`w:line`). O que ela faz é formatar, não converter.
    /// </remarks>
    private static string Invariant(int value) => value.ToString(CultureInfo.InvariantCulture);

    /// <summary>
    /// A medida como o atributo a escreve, ou a **ausência** do atributo.
    /// </summary>
    /// <remarks>
    /// O tipo de retorno é o que importa: devolvendo `StringValue?`, o `null`
    /// apaga o atributo. Escrito como ternário sobre `string`, o mesmo `null`
    /// atravessa a conversão implícita e chega ao SDK como um `StringValue` de
    /// valor nulo, que ele grava como `w:ind w:right=""` — fora do esquema.
    /// </remarks>
    private static StringValue? Twips(int? value) =>
        value is null ? null : new StringValue(Invariant(value.Value));
}
